File: apps/server/src/ai/backends/runpod.ts
import { MAX_AI_RESOLUTION, MAX_DENOISE } from '@brushjam/shared';
import { buildWorkflow, fastProfile, negativePromptActive } from './comfyui.js';
import { AbortedError, BackendHttpError, delay, type AIBackend, type GenerateRequest, type BackendCapabilities } from './types.js';

export interface RunpodOptions {
  endpointId: string;
  apiKey: string;
  checkpoint: string;
  cfg?: number;
  vaeTile?: number;
  fastLora?: string;
  baseUrl?: string;
  /**
   * Overall deadline for one generation, counted from the first byte sent.
   * A cold worker on this endpoint spends ~25 s waiting for a machine and then
   * ~30 s loading a 7 GB checkpoint off the network volume before it samples
   * anything, so anything under ~180 s would abandon perfectly healthy jobs.
   */
  timeoutMs?: number;
  /** Gap between /status polls once the job has gone asynchronous. */
  pollIntervalMs?: number;
}

/** What `/runsync` and `/status/{id}` both answer with. */
interface RunpodJob {
  id?: string;
  status?: string;
  output?: { images?: RunpodImage[]; errors?: string[] };
  error?: string;
  delayTime?: number;
  executionTime?: number;
}

interface RunpodImage {
  filename?: string;
  /** `base64` by default; `s3_url` when the worker is configured to upload. */
  type?: string;
  data?: string;
}

/** Terminal job states. Anything else means "still going". */
const DONE = new Set(['COMPLETED', 'FAILED', 'CANCELLED', 'TIMED_OUT']);

/**
 * The same workflow JSON as the local ComfyUI backend, posted to RunPod's
 * `worker-comfyui` serverless image - transport-only difference, which is the
 * whole point of building the workflow in TS.
 *
 * Verified against a live endpoint (`runpod/worker-comfyui:5.10.0-base` with the
 * checkpoint and LoRA on a network volume) - see docs/RUNPOD.md.
 *
 * Two things about the contract are worth stating, because both were guesses
 * before and only one was right:
 *
 * - Output images arrive as `output.images[].data`, base64, no data-URI prefix.
 * - `/runsync` is not synchronous for long jobs. RunPod gives it about 90 s and
 *   then answers `{id, status: "IN_PROGRESS"}`, leaving the caller to poll
 *   `/status/{id}`. A cold start here is 80-120 s, so that path is the normal
 *   one for the first generation, not an edge case.
 */
export class RunpodBackend implements AIBackend {
  readonly name = 'runpod';

  constructor(private readonly opts: RunpodOptions) {}

  async capabilities(): Promise<BackendCapabilities> {
    // Same workflow builder as ComfyUI, so the same two profiles.
    const cfg = this.opts.cfg ?? 5.5;
    return {
      profiles: this.opts.fastLora ? ['fast', 'quality'] : ['quality'],
      maxResolution: MAX_AI_RESOLUTION,
      maxDenoise: MAX_DENOISE,
      negativePromptActive: {
        fast: negativePromptActive(fastProfile(this.opts.fastLora), cfg),
        quality: negativePromptActive(null, cfg),
      },
    };
  }

  private get base(): string {
    return (this.opts.baseUrl ?? 'https://api.runpod.ai/v2').replace(/\/+$/, '');
  }

  private get url(): string {
    return `${this.base}/${this.opts.endpointId}`;
  }

  private get headers(): Record<string, string> {
    return { 'content-type': 'application/json', authorization: `Bearer ${this.opts.apiKey}` };
  }

  async generate(req: GenerateRequest, signal: AbortSignal): Promise<Buffer> {
    if (signal.aborted) throw new AbortedError();
    const stamp = `${req.tag}_${Date.now()}`;
    const imageName = `brushjam_${stamp}_img.png`;
    const maskName = `brushjam_${stamp}_mask.png`;
    const workflow = buildWorkflow({
      checkpoint: this.opts.checkpoint,
      prompt: req.prompt,
      negativePrompt: req.negativePrompt,
      imageName,
      maskName,
      seed: req.seed,
      steps: req.steps,
      cfg: this.opts.cfg ?? 5.5,
      vaeTile: this.opts.vaeTile ?? 512,
      fastLora: req.profile === 'fast' ? this.opts.fastLora || undefined : undefined,
      denoise: req.denoise,
      filenamePrefix: `brushjam/${req.tag}`,
    });

    const deadline = Date.now() + (this.opts.timeoutMs ?? 300_000);
    const res = await this.post(
      '/runsync',
      {
        input: {
          workflow,
          images: [
            { name: imageName, image: req.imagePng.toString('base64') },
            { name: maskName, image: req.maskPng.toString('base64') },
          ],
        },
      },
      signal,
      deadline,
    );
    if (!res.ok) {
      throw new BackendHttpError(`RunPod /runsync failed: ${res.status} ${(await res.text()).slice(0, 300)}`, res.status);
    }
    let job = (await res.json()) as RunpodJob;

    // /runsync only waits ~90 s. Past that the job is still running on RunPod's
    // side and has to be polled - and abandoning it without a /cancel would
    // leave the single worker busy while the scheduler retries behind it.
    if (!DONE.has(job.status ?? '')) {
      const id = job.id;
      if (!id) throw new Error(`RunPod returned status ${job.status ?? 'unknown'} with no job id`);
      try {
        job = await this.poll(id, signal, deadline);
      } catch (err) {
        await this.cancel(id);
        throw err;
      }
    }
    return decodeImage(job);
  }

  /** One POST, bounded by both the caller's signal and the overall deadline. */
  private async post(path: string, body: unknown, signal: AbortSignal, deadline: number): Promise<Response> {
    return this.call(path, { method: 'POST', headers: this.headers, body: JSON.stringify(body) }, signal, deadline);
  }

  private async call(path: string, init: RequestInit, signal: AbortSignal, deadline: number): Promise<Response> {
    const left = Math.max(1000, deadline - Date.now());
    const timeout = AbortSignal.timeout(left);
    try {
      return await fetch(`${this.url}${path}`, { ...init, signal: AbortSignal.any([signal, timeout]) });
    } catch (err) {
      if (signal.aborted) throw new AbortedError();
      const name = (err as { name?: string })?.name;
      if (name === 'TimeoutError' || name === 'AbortError') throw new Error(`RunPod request timed out: ${path}`);
      throw err;
    }
  }

  private async poll(id: string, signal: AbortSignal, deadline: number): Promise<RunpodJob> {
    const interval = this.opts.pollIntervalMs ?? 1000;
    let lastPollError: string | null = null;
    for (;;) {
      if (signal.aborted) throw new AbortedError();
      try {
        const res = await this.call(`/status/${encodeURIComponent(id)}`, { headers: this.headers }, signal, deadline);
        if (res.ok) {
          const job = (await res.json()) as RunpodJob;
          if (DONE.has(job.status ?? '')) return job;
          lastPollError = null;
        } else if (res.status >= 500 || res.status === 429) {
          // The job outlives a blip on the control plane; keep polling.
          lastPollError = `status ${res.status}`;
        } else {
          throw new BackendHttpError(`RunPod /status failed: ${res.status}`, res.status);
        }
      } catch (err) {
        if (err instanceof AbortedError || err instanceof BackendHttpError) throw err;
        lastPollError = err instanceof Error ? err.message : String(err);
      }
      if (Date.now() > deadline) {
        throw new Error(lastPollError ? `RunPod generation timed out (last poll: ${lastPollError})` : 'RunPod generation timed out');
      }
      await delay(interval, signal);
    }
  }

  /** Best effort: a job left running would block the endpoint's only worker. */
  private async cancel(id: string): Promise<void> {
    try {
      await fetch(`${this.url}/cancel/${encodeURIComponent(id)}`, {
        method: 'POST',
        headers: this.headers,
        signal: AbortSignal.timeout(5000),
      });
    } catch {
      /* the result is dropped either way */
    }
  }
}

/**
 * Pull the PNG out of a finished job. Kept separate so the smoke script and the
 * tests can reuse the exact error messages the server would print.
 */
export function decodeImage(job: RunpodJob): Buffer {
  if (job.status && job.status !== 'COMPLETED') {
    const detail = job.error ?? job.output?.errors?.join('; ') ?? '';
    throw new Error(`RunPod job ${job.status}${detail ? `: ${detail}` : ''}`);
  }
  if (job.error) throw new Error(`RunPod error: ${job.error}`);
  const first = job.output?.images?.[0];
  if (!first) {
    const detail = job.output?.errors?.join('; ');
    throw new Error(detail ? `RunPod returned no image: ${detail}` : 'RunPod returned no image');
  }
  if (first.type === 's3_url') {
    throw new Error('RunPod worker is configured for S3 upload; this backend expects base64 images');
  }
  if (!first.data) throw new Error('RunPod returned an image entry with no data');
  return Buffer.from(first.data.replace(/^data:image\/\w+;base64,/, ''), 'base64');
}

File: apps/brushjam/src/brushjam/ai/backends/runpod.py
"""RunPod serverless backend. Port of the retired Node server's src/ai/backends/runpod.ts.

The same workflow JSON as the local ComfyUI backend, posted to RunPod's
`worker-comfyui` image, so moving to the cloud is a transport change only.

`/runsync` is not synchronous for long jobs: RunPod gives it about 90 s and
then answers `{id, status: "IN_PROGRESS"}`, leaving the caller to poll
`/status/{id}`. A cold start here is 80-120 s, so that path is the normal one
for the first generation, not an edge case.
"""

from __future__ import annotations

import asyncio
import base64
import re
import time
from typing import Any, Dict, Optional

import httpx

from ...constants import MAX_AI_RESOLUTION, MAX_DENOISE
from .base import BackendCapabilities, BackendHttpError, GenerateRequest
from .comfyui import build_workflow, fast_profile, negative_prompt_active

#: Terminal job states. Anything else means "still going".
DONE = {"COMPLETED", "FAILED", "CANCELLED", "TIMED_OUT"}


class RunpodBackend:
    name = "runpod"

    def __init__(
        self,
        endpoint_id: str,
        api_key: str,
        checkpoint: str,
        cfg: float = 5.5,
        vae_tile: int = 512,
        fast_lora: Optional[str] = None,
        base_url: str = "https://api.runpod.ai/v2",
        timeout_ms: float = 300_000,
        poll_interval_ms: float = 1000,
        transport: Any = None,
    ) -> None:
        self.endpoint_id = endpoint_id
        self.api_key = api_key
        self.checkpoint = checkpoint
        self.cfg = cfg
        self.vae_tile = vae_tile
        self.fast_lora = fast_lora or None
        self.base_url = base_url.rstrip("/")
        self.timeout_ms = timeout_ms
        self.poll_interval_ms = poll_interval_ms
        #: Test seam: an httpx transport to use instead of the network.
        self._transport = transport

    @property
    def url(self) -> str:
        return f"{self.base_url}/{self.endpoint_id}"

    @property
    def headers(self) -> Dict[str, str]:
        return {"content-type": "application/json", "authorization": f"Bearer {self.api_key}"}

    async def capabilities(self) -> BackendCapabilities:
        # Same workflow builder as ComfyUI, so the same two profiles.
        return BackendCapabilities(
            profiles=["fast", "quality"] if self.fast_lora else ["quality"],
            max_resolution=MAX_AI_RESOLUTION,
            max_denoise=MAX_DENOISE,
            negative_prompt_active={
                "fast": negative_prompt_active(fast_profile(self.fast_lora), self.cfg),
                "quality": negative_prompt_active(None, self.cfg),
            },
        )

    async def generate(self, req: GenerateRequest) -> bytes:
        stamp = f"{req.tag}_{int(time.time() * 1000)}"
        image_name = f"brushjam_{stamp}_img.png"
        mask_name = f"brushjam_{stamp}_mask.png"
        workflow = build_workflow(
            checkpoint=self.checkpoint,
            prompt=req.prompt,
            negative_prompt=req.negative_prompt,
            image_name=image_name,
            mask_name=mask_name,
            seed=req.seed,
            steps=req.steps,
            cfg=self.cfg,
            vae_tile=self.vae_tile,
            fast_lora=self.fast_lora if req.profile == "fast" else None,
            denoise=req.denoise,
            filename_prefix=f"brushjam/{req.tag}",
        )
        deadline = time.monotonic() + self.timeout_ms / 1000
        async with httpx.AsyncClient(
            base_url=self.url, headers=self.headers, transport=self._transport
        ) as client:
            res = await self._call(
                client,
                "POST",
                "/runsync",
                deadline,
                json={
                    "input": {
                        "workflow": workflow,
                        "images": [
                            {"name": image_name, "image": base64.b64encode(req.image_png).decode("ascii")},
                            {"name": mask_name, "image": base64.b64encode(req.mask_png).decode("ascii")},
                        ],
                    }
                },
            )
            if res.status_code >= 400:
                raise BackendHttpError(
                    f"RunPod /runsync failed: {res.status_code} {res.text[:300]}", res.status_code
                )
            job = res.json()
            if job.get("status") not in DONE:
                job_id = job.get("id")
                if not job_id:
                    raise RuntimeError(
                        f"RunPod returned status {job.get('status') or 'unknown'} with no job id"
                    )
                try:
                    job = await self._poll(client, job_id, deadline)
                except BaseException:
                    await self._cancel(job_id)
                    raise
        return decode_image(job)

    async def _call(
        self, client: httpx.AsyncClient, method: str, path: str, deadline: float, **kwargs: Any
    ) -> httpx.Response:
        left = max(1.0, deadline - time.monotonic())
        try:
            return await client.request(method, path, timeout=left, **kwargs)
        except httpx.TimeoutException as err:
            raise RuntimeError(f"RunPod request timed out: {path}") from err

    async def _poll(self, client: httpx.AsyncClient, job_id: str, deadline: float) -> Dict[str, Any]:
        last_poll_error: Optional[str] = None
        while True:
            try:
                res = await self._call(client, "GET", f"/status/{job_id}", deadline)
                if res.status_code < 400:
                    job = res.json()
                    if job.get("status") in DONE:
                        return job
                    last_poll_error = None
                elif res.status_code >= 500 or res.status_code == 429:
                    # The job outlives a blip on the control plane; keep polling.
                    last_poll_error = f"status {res.status_code}"
                else:
                    raise BackendHttpError(f"RunPod /status failed: {res.status_code}", res.status_code)
            except (asyncio.CancelledError, BackendHttpError):
                raise
            except Exception as err:  # noqa: BLE001
                last_poll_error = str(err)
            if time.monotonic() > deadline:
                raise RuntimeError(
                    f"RunPod generation timed out (last poll: {last_poll_error})"
                    if last_poll_error
                    else "RunPod generation timed out"
                )
            await asyncio.sleep(self.poll_interval_ms / 1000)

    async def _cancel(self, job_id: str) -> None:
        """Best effort: a job left running would block the endpoint's only worker."""
        try:
            async with httpx.AsyncClient(
                base_url=self.url, headers=self.headers, timeout=5, transport=self._transport
            ) as client:
                await client.post(f"/cancel/{job_id}")
        except Exception:  # noqa: BLE001
            pass


_DATA_URL = re.compile(r"^data:image/\w+;base64,")


def decode_image(job: Dict[str, Any]) -> bytes:
    """Pull the PNG out of a finished job, with the exact error messages the
    server would print."""
    status = job.get("status")
    if status and status != "COMPLETED":
        detail = job.get("error") or "; ".join((job.get("output") or {}).get("errors") or [])
        raise RuntimeError(f"RunPod job {status}{': ' + detail if detail else ''}")
    if job.get("error"):
        raise RuntimeError(f"RunPod error: {job['error']}")
    images = (job.get("output") or {}).get("images") or []
    if not images:
        detail = "; ".join((job.get("output") or {}).get("errors") or [])
        raise RuntimeError(f"RunPod returned no image: {detail}" if detail else "RunPod returned no image")
    first = images[0]
    if first.get("type") == "s3_url":
        raise RuntimeError(
            "RunPod worker is configured for S3 upload; this backend expects base64 images"
        )
    if not first.get("data"):
        raise RuntimeError("RunPod returned an image entry with no data")
    return base64.b64decode(_DATA_URL.sub("", first["data"]))
